integrationtest/mapper/orm-spring/src/test/java/org/hibernate/search/integrationtest/spring/beanresolution/SpringBeanResolutionIT.java

```diff
@@ -0,0 +1,277 @@
+		}
+	}
+
+	@Rule
+	public BackendMock backendMock = new BackendMock();
+
+	@Rule
+	public OrmSetupHelper ormSetupHelper = OrmSetupHelper.withBackendMock( backendMock );
+
+	@Rule
+	public StaticCounters counters = new StaticCounters();
+
+	@Test
+	public void singleton_byType() {
+		doTest(
+				ExpectedScope.SINGLETON, UnnamedSingletonBean.KEYS,
+				BeanReference.of( UnnamedSingletonBean.class )
+		);
+	}
+
+	@Test
+	public void singleton_byName() {
+		doTest(
+				ExpectedScope.SINGLETON, NamedSingletonBean.KEYS,
+				BeanReference.of( InterfaceDefinedByMapper.class, NamedSingletonBean.NAME )
+		);
+	}
+
+	@Test
+	public void prototype_byType() {
+		doTest(
+				ExpectedScope.PROTOTYPE, UnnamedPrototypeBean.KEYS,
+				BeanReference.of( UnnamedPrototypeBean.class )
+		);
+	}
+
+	@Test
+	public void prototype_byName() {
+		doTest(
+				ExpectedScope.PROTOTYPE, NamedPrototypeBean.KEYS,
+				BeanReference.of( InterfaceDefinedByMapper.class, NamedPrototypeBean.NAME )
+		);
+	}
+
+	private <T> void doTest(ExpectedScope expectedScope, CounterKeys counterKeys, BeanReference<T> reference) {
+		List<BeanHolder<T>> retrievedBeans = new ArrayList<>();
+
+		backendMock.onCreate( context -> {
+			BeanHolder<T> retrievedBean1 = context.beanResolver().resolve( reference );
+			retrievedBeans.add( retrievedBean1 );
+			BeanHolder<T> retrievedBean2 = context.beanResolver().resolve( reference );
+			retrievedBeans.add( retrievedBean2 );
+		} );
+		backendMock.onStop( () -> {
+			for ( BeanHolder<T> retrievedBean : retrievedBeans ) {
+				retrievedBean.close();
+			}
+		} );
+		backendMock.expectAnySchema( IndexedEntity.NAME );
+
+		int expectedInstances = ExpectedScope.SINGLETON.equals( expectedScope ) ? 1 : 2;
+
+		try ( @SuppressWarnings("unused") ConfigurableApplicationContext applicationContext = startApplication() ) {
+			applicationContext.getBean( EntityManagerFactory.class ).getMetamodel();
+			backendMock.verifyExpectationsMet();
+
+			assertThat( retrievedBeans )
+					.hasSize( 2 )
+					.doesNotContainNull()
+					.extracting( BeanHolder::get )
+					.doesNotContainNull()
+					// All created beans should have been injected
+					.allSatisfy( bean -> assertThat( bean )
+							.asInstanceOf( InstanceOfAssertFactories.type( AbstractBeanBase.class ) )
+							.extracting( "injectedBean" )
+							.isNotNull()
+					);
+
+			assertThat( StaticCounters.get().get( counterKeys.instantiate ) ).isEqualTo( expectedInstances );
+			assertThat( StaticCounters.get().get( counterKeys.postConstruct ) ).isEqualTo( expectedInstances );
+			assertThat( StaticCounters.get().get( counterKeys.preDestroy ) ).isEqualTo( 0 );
+		}
+
+		switch ( expectedScope ) {
+			case PROTOTYPE:
+				// Spring does not call @PreDestroy on prototype beans
+				// See https://docs.spring.io/spring-framework/docs/5.3.1/reference/html/core.html#beans-factory-scopes-prototype
+				assertThat( StaticCounters.get().get( counterKeys.preDestroy ) ).isEqualTo( 0 );
```

Comment:
Maybe if the framework expected behavior is not respected, we should assert nothing about it.

Reply:
I'd like to know if they change the behavior.